src/pytest/__init__.py

```diff
@@ -46,8 +46,57 @@
 from _pytest.warning_types import PytestUnknownMarkWarning
 from _pytest.warning_types import PytestWarning
 
+# For mypy Any type checking purposes.
+# This file sets disallow_any_expr to ensure that the public API
+# does not have dynamic typing via Any. Manually using each public API
+# type as an expression to enforce this.
+__version__ = __version__
+register_assert_rewrite = register_assert_rewrite
+_setup_collect_fakemodule = _setup_collect_fakemodule
+cmdline = cmdline
+ExitCode = ExitCode
+#hookimpl = hookimpl
+#hookspec = hookspec
```

Comment:
hookimpl and hookspec are the two which aren't very easy to type.
I believe we want to do something similar to #3342 for pluggy in order to make this possible.

Reply:
With regard to hookimpl/hookspec it would also be great to have some plugin or similar that would make mypy aware of hookimpls beloning to hookspecs of the same name.
(Maybe also / additionally separate types could be used, but that might be tedious)
(Just throwing in some food for thought)

Reply:
The hookimpl and hookspec decorator can be typed relatively easily - but in order to do so, we need to have type stubs or exported types from pluggy.

Reply:
I think it is more complex: how does mypy know that `@hookimpl(pytest_foo)` refers to `@hookspec(pytest_foo)`?
What I've meant here is that the former should be validated against the latter somehow.

Reply:
https://github.com/pytest-dev/pluggy/blob/master/src/pluggy/hooks.py#L22
HookSpec looks like it is a decorator which takes a handful of bools as arguments

I suppose the project_name is something which we could encode into the type system (`HookSpecMarker[_T]` where _T is the project name).

I'm not sure I follow how hookimpl and hookspec are tied together here.